js-examples/addressbook.js

```diff
@@ -0,0 +1,142 @@
+  Person_PhoneNumber_Type
+} = require('./addressbook.capnp.js');
+
+function writePackedMessageToStream(writeStream, message) {
+  const arrayBuffer = message.toPackedArrayBuffer();
+  // Beacause streams can't handle ArrayBuffers
+  const buffer = new Buffer(arrayBuffer);
+  writeStream.write(buffer);
+}
+
+function writeAddressBook(writeStream) {
+  const message = new capnp.Message();
+  const addressBook = message.initRoot(AddressBook);
+  const people = addressBook.initPeople(2);
+
+  const alice = people.get(0);
+  alice.setId(123);
+  alice.setName("Alice");
+  alice.setEmail("alice@example.com");
+  const alicePhones = alice.initPhones(1);
+  alicePhones.get(0).setNumber("555-1212");
+  alicePhones.get(0).setType(Person_PhoneNumber_Type.MOBILE);
+  alice.getEmployment().setSchool("MIT");
+
+  const bob = people.get(1);
+  bob.setId(456);
+  bob.setName("Bob");
+  bob.setEmail("bob@example.com");
+  const bobPhones = bob.initPhones(2);
+  bobPhones.get(0).setNumber("555-4567");
+  bobPhones.get(0).setType(Person_PhoneNumber_Type.HOME);
+  bobPhones.get(1).setNumber("555-7654");
+  bobPhones.get(1).setType(Person_PhoneNumber_Type.WORK);
+  bob.getEmployment().setUnemployed();
+
+  writePackedMessageToStream(writeStream, message);
+}
+
+function readToEndOfStream(readStream) {
+  return new Promise(function(resolve, reject) {
+    let result = new Uint8Array();
+    readStream.on('data', function(data) {
+      const oldLen = result.byteLength;
+      const newResult = new Uint8Array(oldLen + data.byteLength);
+      newResult.set(result);
+      newResult.set(data, oldLen);
+      result = newResult;
+    });
+    readStream.on('end', function() {
+      resolve(result);
+    });
+    readStream.on('error', reject);
+  });
+}
+
+function printAddressBook(readStream) {
+  return readToEndOfStream(readStream).then(function(data) {
+    const message = capnp.Message.fromPackedArrayBuffer(data);
+    const addressBook = message.getRoot(AddressBook);
+    addressBook.getPeople().forEach(function(person) {
+      console.log(person.getName() + ': ' + person.getEmail());
+      person.getPhones().forEach(function(phone) {
+        let phoneTypeString = Person_PhoneNumber_Type[phone.getType()];
+        console.log('  ' + phoneTypeString + ' phone: ' + phone.getNumber());
+      });
+      const employment = person.getEmployment();
+      switch (employment.which()) {
+        case Person_Employment.UNEMPLOYED:
+          console.log('  unemployed');
+          break;
+        case Person_Employment.EMPLOYER:
+          console.log('  employer: ' + employment.getEmployer());
+          break;
+        case Person_Employment.SCHOOL:
+          console.log('  student at: ' + employment.getSchool());
+          break;
+        case Person_Employment.SELF_EMPLOYED:
+          console.log('  self-employed');
+          break;
+      }
+    });
+  });
+}
+
+function main() {
+  // TODO Not sure if we have dynamic schema usage available yet
```

Comment:
If you have a look at the C++ example https://github.com/capnproto/capnproto/blob/master/c%2B%2B/samples/addressbook.c%2B%2B#L264 it appears that here "dynamic" doesn't mean reading the schema at runtime. Rather I think it just pulls out the raw schema's from the compiled struct. eg. The member at: https://github.com/capnproto/capnproto/blob/cff543590ab25652ce56e9178dd499a8a69e08f7/c%2B%2B/src/capnp/schema.capnp.c%2B%2B#L689
so it can use it in a "dynamic" fashion. These schema's seem somewhat equivalent to our ObjectSize objects, but carry a bit more information.

Obviously in JavaScript we _could_ just do things like:
```
function setFieldInCapnpStruct ( struct, fieldname, value ) {
    // ignore my lack of attention to the casing
    struct['set' + fieldname ](value);
}

function printAllCapnpFields ( struct ) {
    for(var prop in struct) {
        // I'm improvising here, my javascript is patchy, but you should get the idea
        if (/*has own property etc*/) {
            if(prop.startsWith('get') && struct[prop] instanceof Function) {
                console.log(prop.replace('get', '') + ' = ' + struct[prop].apply(struct));
            }
        }
    }
}
```
Which means having a dynamic API to the capnp data does not yield an increase in value anywhere near the same as in C++ with its meagre reflection tools out of the box.

However, I think it might still be preferable to support these api's if we can. As you'll still receive much better type information when doing things similar to the above example, and they'd be even simpler and more performant than what I've shown above. Also, note what I did above was super hacky and doesn't handle anywhere near enough edge cases which wouldn't even be a concern if you had the first class type information from capnp, so if someone tried the above in lieu of access to the compiled schema's it would be harder still.

Might be good to get @kentonv 's thoughts here too? What do you think about dynamic api support that doesn't involve parsing schema's at runtime?